gemstone/src/swapper/mayan/swift_provider.rs

```diff
@@ -0,0 +1,559 @@
+        let gas_drop = self.convert_amount_to_wei(quote.gas_drop, quote.to_token.decimals.into()).map_err(|e| {
+            eprintln!("Failed to convert gas_drop: {}", quote.gas_drop);
+            e
+        })?;
+
+        let random_bytes = Self::generate_random_bytes32();
+
+        let referrer = self.get_referrer(request)?;
+
+        let params = OrderParams {
+            trader: trader_address,
+            token_out,
+            min_amount_out: min_amount_out.parse().map_err(|_| SwapperError::InvalidAmount)?,
+            gas_drop: gas_drop.parse().map_err(|_| SwapperError::InvalidAmount)?,
+            cancel_fee: quote.cancel_relayer_fee64.parse::<u64>().map_err(|_| SwapperError::InvalidAmount)?,
+            refund_fee: quote.refund_relayer_fee64.parse::<u64>().map_err(|_| SwapperError::InvalidAmount)?,
+            deadline,
+            dest_addr: destination_address,
+            dest_chain_id: dest_chain_id.to_string().parse().map_err(|_| SwapperError::InvalidAmount)?,
+            referrer_addr: referrer.clone().map_or([0u8; 32], |x| {
+                self.to_native_wormhole_address(x.address.as_str(), from_chain_id)
+                    .map_err(|_| SwapperError::ComputeQuoteError {
+                        msg: "Unable to get referrer wormhole address".to_string(),
+                    })
+                    .unwrap()
+            }),
+            referrer_bps: referrer.map_or(0u8, |x| x.bps.try_into().map_err(|_| SwapperError::InvalidAmount).unwrap()),
+            auction_mode: quote.swift_auction_mode.unwrap_or(0),
+            random: random_bytes,
+        };
+
+        Ok(params)
+    }
+
+    fn generate_random_bytes32() -> [u8; 32] {
+        let mut rng = rand::thread_rng();
+        let mut random_bytes = [0u8; 32];
+        rng.fill(&mut random_bytes);
+        random_bytes
+    }
+
+    fn address_to_bytes32(&self, address: &str) -> Result<[u8; 32], SwapperError> {
+        let addr = EthereumAddress::from_str(address).map_err(|_| SwapperError::InvalidAddress { address: address.to_string() })?;
+        let mut bytes32 = [0u8; 32];
+        bytes32[12..].copy_from_slice(&addr.bytes);
+        Ok(bytes32)
+    }
+
+    fn get_referrer(&self, request: &SwapQuoteRequest) -> Result<Option<SwapReferralFee>, SwapperError> {
+        if let Some(options) = &request.options {
+            if let Some(referrer) = &options.fee {
+                let evm_fee = &referrer.evm;
+                let solana_fee = &referrer.solana;
+
+                if request.from_asset.chain == Chain::Solana {
+                    return Ok(Some(solana_fee.clone()));
+                }
+
+                return Ok(Some(evm_fee.clone()));
+            }
+        }
+
+        Ok(None)
+    }
+
+    async fn fetch_quote_from_request(&self, request: &SwapQuoteRequest, provider: Arc<dyn AlienProvider>) -> Result<Quote, SwapperError> {
+        let asset_decimals = self.get_asset_decimals(request.from_asset.clone(), provider.clone()).await?;
```

Comment:
@gemcoder21 let's pass decimal in for assets? to save a network request, or we can add `X-Cache-TTL` for this request, wdyt?

Reply:
Yes, that's a good idea, I know it's used in other places too, like Thorchain. So pass `Asset` instead of `AssetId` in swap quote request?